File: IOS UIDynamics-Animation/AnimatorDemo/ViewController.m
//
//  ViewController.m
//  AnimatorDemo
//
//  Created by Prateek Sharma on 7/4/17.
//  Copyright © 2017 Bingo Tech. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIView *viewInner;
@property (strong, nonatomic) UIDynamicAnimator *animator;
@property (strong, nonatomic) UIGravityBehavior *gravity;
@property (strong, nonatomic) UICollisionBehavior *collider;
@end

@implementation ViewController

- (UIDynamicAnimator *)animator{
    
    if(!_animator){
        _animator = [[UIDynamicAnimator alloc] initWithReferenceView:self.viewInner];
    }
    
    return _animator;
}

- (UIGravityBehavior *)gravity{
    if(!_gravity){
        _gravity = [[UIGravityBehavior alloc] init];
        _gravity.magnitude=0.9;
        
        [self.animator addBehavior:_gravity];
    }
    return _gravity;
}

- (UICollisionBehavior *)collider{
    if(!_collider){
        _collider = [[UICollisionBehavior alloc] init];
        
        /* THis will make the collider use the bounds of the reference view as the bound for collision */
        
        _collider.translatesReferenceBoundsIntoBoundary = YES;
        
        [self.animator addBehavior:_collider];
        
    }
    return _collider;
}


static const CGSize DROP_PT = {40,40};

- (IBAction)tapAction1:(UITapGestureRecognizer *)sender {
    [self drop];
}

- (void)drop{

    /* This will basically make a block frame of size 40x40
        on the upper side of the mobile .
     Thats why , the frame is given an origin of 0.
     The width is then managed by gettinig a ranadom number,
     bringing it inside the window width value, then finding the point to show that frame by dividing it with the frame width and getting an absolute floor value.
     
     Then just getting a random color and adding the view aas a subview in our main view.
     
     */
    
    
    CGRect frame;
    frame.origin = CGPointZero;
    frame.size = DROP_PT;
    int x = (arc4random()%(int)self.viewInner.bounds.size.width) / DROP_PT.width;
    
    frame.origin.x = x * DROP_PT.width;
    
    
    UIView *dropView = [[UIView alloc] initWithFrame:frame];
    dropView.backgroundColor = [self randomColor];
    [self.viewInner addSubview:dropView];
    
    
    //This will add the view in the gravity animation part
    
    [self.gravity addItem:dropView];
    [self.collider addItem:dropView];
    
    
}

- (UIColor *)randomColor{
    
    //UIColor *color;
    
    switch((arc4random()%5)){
        case 0: return [UIColor blueColor];
        case 1: return [UIColor magentaColor];
        case 2: return [UIColor orangeColor];
        case 3: return [UIColor cyanColor];
        case 4: return [UIColor greenColor];
        default: return [UIColor brownColor];
    }
    
    return [UIColor blackColor];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: PopoutAnimation/ViewController.swift
//
//  ViewController.swift
//  PopoutAnimation
//
//  Created by Prateek on 10/09/17.
//  Copyright © 2017 14K. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var mainBtn: UIButton!
    @IBOutlet weak var btn1: UIButton!
    @IBOutlet weak var btn2: UIButton!
    @IBOutlet weak var btn3: UIButton!
    
    var btn1Center : CGPoint!
    var btn2Center : CGPoint!
    var btn3Center : CGPoint!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        btn1Center = btn1.center
        btn2Center = btn2.center
        btn3Center = btn3.center
        
        btn1.center = mainBtn.center;
        btn2.center = mainBtn.center;
        btn3.center = mainBtn.center;

        btn1.alpha = 0.0
        btn2.alpha = 0.0
        btn3.alpha = 0.0
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func mainAction(_ sender: UIButton) {
        
        if(sender.tag == 0){
            sender.backgroundColor = UIColor.orange
            sender.tag = 1;
        
            
            UIView.animate(withDuration: 0.3) {
                
                self.btn1.alpha = 1.0
                self.btn1.center = self.btn1Center
                self.btn2.alpha = 1.0
                self.btn2.center = self.btn2Center
                self.btn3.alpha = 1.0
                self.btn3.center = self.btn3Center
            }
        
        
        }
        else{
            sender.backgroundColor = UIColor.clear
            sender.tag = 0;
            
            
            UIView.animate(withDuration: 0.3) {
                
                self.btn1.alpha = 0.0
                self.btn1.center = self.mainBtn.center
                self.btn2.alpha = 0.0
                self.btn2.center = self.mainBtn.center
                self.btn3.alpha = 0.0
                self.btn3.center = self.mainBtn.center
            }
            
            
        }
        
        
        
    }
    
    

    @IBAction func btn1Action(_ sender: UIButton) {
    
        
        if(sender.tag == 0){
            sender.backgroundColor = UIColor.orange
            sender.tag = 1;
        }
        else{
            sender.backgroundColor = UIColor.clear
            sender.tag = 0;
        }
        
    }
    @IBAction func btn2Action(_ sender: UIButton) {
        
        
        if(sender.tag == 0){
            sender.backgroundColor = UIColor.orange
            sender.tag = 1;
        }
        else{
            sender.backgroundColor = UIColor.clear
            sender.tag = 0;
        }
        
    }
    @IBAction func btn3Action(_ sender: UIButton) {
    
        
        if(sender.tag == 0){
            sender.backgroundColor = UIColor.orange
            sender.tag = 1;
        }
        else{
            sender.backgroundColor = UIColor.clear
            sender.tag = 0;
        }
    
    }
    
    
}

